src/internal_modules/roc_packet/delayed_reader.cpp

```diff
@@ -23,24 +24,25 @@ DelayedReader::DelayedReader(IReader& reader,
     roc_log(LogDebug, "delayed reader: initializing: delay=%lu", (unsigned long)delay_);
 }
 
-PacketPtr DelayedReader::read() {
+int DelayedReader::read(PacketPtr& ptr) {
     if (!started_) {
         if (!fetch_packets_()) {
-            return NULL;
+            return error::ErrUnknown;
         }
 
         started_ = true;
     }
 
     if (queue_.size() != 0) {
-        return read_queued_packet_();
+        return read_queued_packet_(ptr);
     }
 
-    return reader_.read();
+    return reader_.read(ptr);
 }
 
 bool DelayedReader::fetch_packets_() {
-    while (PacketPtr pp = reader_.read()) {
+    PacketPtr pp;
+    while (reader_.read(pp) == 0) {
         queue_.write(pp);
     }
```

Comment:
We should handle returned code:

* if code is ErrNoMore, it's OK, we should just return
* if it's another code, it's failure, we should forward it to upper levels